git/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+    pub target: Target,
+}
+
+impl Spec {
+    /// The HEAD commit at the given url.
+    pub fn head(url: gix::Url) -> Self {
+        Self {
+            url,
+            target: Target::Head,
+        }
+    }
+
+    pub fn commit(url: gix::Url, commit: ObjectId) -> Self {
+        Self {
+            url,
+            target: Target::Commit(commit),
+        }
+    }
+}
+
+/// The different kinds of git "thing" that we can target.
+#[serde_with::serde_as]
+#[derive(Clone, Debug, PartialEq, Eq, Hash, Default, serde::Serialize, serde::Deserialize)]
+pub enum Target {
+    /// By default, we target the remote HEAD.
+    #[default]
+    Head,
+    /// Target the tip of a specific branch.
+    Branch(String),
+    /// Target a specific tag.
+    Tag(String),
+    /// Target a specific commit.
+    ///
+    /// Currently, we only support a full commit: this needs to be a full hex-encoded
+    /// sha hash. We could try to support prefixes also, but it requires some work
+    /// and it *appears* to be inherently less efficient because there doesn't seem
+    /// to be a way to fetch it in one shot. At least, when `cargo` needs to fetch an
+    /// abbreviated hash it fetches everything and then looks for the hash among the
+    /// things that it finds.
+    Commit(#[serde_as(as = "serde_with::DisplayFromStr")] ObjectId),
+}
+
+impl std::fmt::Display for Target {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Target::Head => write!(f, "HEAD"),
+            Target::Branch(branch) => write!(f, "refs/heads/{branch}"),
+            Target::Tag(tag) => write!(f, "refs/tags/{tag}"),
+            Target::Commit(c) => write!(f, "{}", c),
+        }
+    }
+}
+
+impl Target {
+    fn refspec(&self) -> String {
+        self.to_string()
+    }
+}
+
+fn source_object_id(source: &gix::remote::fetch::Source) -> Result<ObjectId> {
+    match source {
+        gix::remote::fetch::Source::ObjectId(id) => Ok(*id),
+        gix::remote::fetch::Source::Ref(r) => {
+            let (_name, id, peeled) = r.unpack();
+
+            Ok(peeled
+                .or(id)
+                .ok_or_else(|| anyhow!("unborn reference"))?
+                .to_owned())
+        }
+    }
+}
+
+/// Fetches the contents of a git repository into directory.
+///
+/// The directory will be created if it doesn't exist yet. The data will be
+/// fetched from scratch, even if it has already been fetched before. However,
+/// we will try to minimize the amount of data to fetch (for example, by doing a
+/// shallow fetch).
+///
+/// Only the contents of the git repository will be written to the given
+/// directory; the git directory itself will be discarded.
+pub fn fetch(spec: &Spec, dir: impl AsRef<Path>) -> Result<ObjectId> {
+    let dir = dir.as_ref();
+    std::fs::create_dir_all(dir).with_path(dir)?;
+
+    // Fetch the git directory somewhere temporary.
+    let git_tempdir = tempfile::tempdir().wrap_err()?;
```

Comment:
Maybe a naive question, but why do we need a temporary directory first? Can't we do everything in `dir` directly?

Reply:
We need to put the `.git` stuff somewhere. There's probably no harm in putting it in `dir`, but I thought it was cleaner to just populate `dir` with the contents, and not the `.git` stuff.

Reply:
Oh, you can choose to put the git stuff separately, I see. That's fine by me :+1: